src/Network/Transport/TCP.hs

```diff
@@ -1450,7 +1452,7 @@ schedule theirEndPoint act = do
 -- 'runScheduledAction' or by another).
 runScheduledAction :: EndPointPair -> Action a -> IO a
 runScheduledAction (ourEndPoint, theirEndPoint) mvar = do
-    join $ readChan (remoteScheduled theirEndPoint)
+    join $ uninterruptibleMask_ $ readChan (remoteScheduled theirEndPoint)
```

Comment:
In order to exploit problematic behaviour you need 2 readers:

```
main = do
      chan <- newChan
      tid1 <- forkIO $ mask_ $ forever $ writeChan chan () >> readChan chan
      tid2 <- forkIO $ mask_ $ forever $ writeChan chan () >> readChan chan
      putStrLn "start"
      killThread tid1
      putStrLn "stop"
```

this program does terminate, so shows the problem.